glambie/util/timeseries_helpers.py

```diff
@@ -0,0 +1,358 @@
+from matplotlib import pyplot as plt
+import numpy as np
+import math
+from scipy import interpolate
+from typing import Union
+
+
+def contains_duplicates(x: np.array) -> bool:
+    """Checks if an array / list contains duplicate values
+    """
+    return len(np.unique(x)) != len(x)
+
+
+def get_matched_indices(array1: np.array, array2: np.array, tolerance: float = 0.0) -> Union[np.array, np.array]:
+    """
+    Returns two arrays of indices at which 'array1' and 'array2' match.
+    Can e.g. be used to find matching dates from two timeseries
+
+    Example
+    -------
+    >>> a = np.array([3, 5, 7, 9, 11])
+    >>> b = np.array([5, 6, 7, 8, 9, 10])
+    >>> ai, bi = get_matched_indices(a, b)
+    >>> print(ai)
+    [1, 2, 3]
+    >>> print(bi)
+    [0, 2, 4]
+    >>> print(a[ai])
+    [5, 7, 9]
+    >>> print(b[bi])
+    [5, 7, 9]
+
+    Parameters
+    ----------
+    array1 : np.array
+        the first of the two series to match
+    array2 : np.array
+        the second of the two series to match
+    tolerance : float, optional
+        the tolerance within which to consider a pair of values to be a match, by default 0.0
+
+    Returns
+    -------
+    Union[np.array, np.array]
+        1. The indices at which values in 'array1' match a value in 'array2'
+        2. The indices at which values in 'array2' match a value in 'array1'
+    """
+    if contains_duplicates(array1) or contains_duplicates(array2):
+        raise ValueError('Input array <array1> or <array2> should not contain duplicates.')
+
+    length_arr1 = len(array1)
+    length_arr2 = len(array2)
+    match_arr1 = np.zeros(array1.shape, dtype=bool)
+    match_arr2 = np.zeros(array2.shape, dtype=bool)
+
+    if length_arr1 == 1 or length_arr2 == 1:
+        if length_arr2 > 1:
+            match_arr2[:] = (array2 == array1[0])
+            if match_arr2.any():
+                match_arr1 = np.zeros((np.count_nonzero(match_arr2)), dtype=np.int32)
+            else:
+                match_arr1 = np.array([])
+            match_arr2 = np.flatnonzero(match_arr2)
+        else:
+            match_arr1[:] = (array1 == array2[0])
+            if match_arr1.any():
+                match_arr2 = np.zeros((np.count_nonzero(match_arr1)), dtype=np.int32)
+            else:
+                match_arr2 = np.array([])
+            match_arr1 = np.flatnonzero(match_arr1)
+        return match_arr1, match_arr2
+    concat_arr = np.concatenate((array1, array2))
+    ind = np.concatenate(
+        [np.arange(length_arr1), np.arange(length_arr2)]
+    )
+    vec = np.concatenate(
+        [np.zeros([length_arr1], dtype=bool),
+         np.ones([length_arr2], dtype=bool)]
+    )
+
+    sub = np.argsort(concat_arr)
```

Comment:
Ah, neat, I see now. 

I can just about follow this, but it would benefit from a few comments within the algorithm itself.

Reply:
hope its a bit more clear now after comments and some variable name changes